### PR TITLE
[PrismNet][ROCm] Return empty grads from the CK flash-attention backward on an empty batch (#195813)

### PR DESCRIPTION
**Pull request by rmiao**
Summary:

`mha_bwd_ck` and `mha_varlen_bwd_ck` carry the same
`TORCH_CHECK(batch_size > 0, "batch size must be positive")` as the forward entries, so
once the forward returns the empty result a training step in which a segment receives no
rows still aborts on the way back. CUDA already returns empty grads here, at
`flash_api.cpp:904`.

The fix takes the same shape as the forward diff below it: drop the check and extend the
"nothing to compute" branch both files already have for an empty query sequence. At zero
rows the grads allocated above that branch are already correctly empty, so only the kernel
launch is skipped.

On a local MI300X a zero-row backward now produces empty grads of the right shape for
query, key and value instead of aborting. One related edge is left alone -- 
`mha_bwd_ck.hip:375` rejects a stride-0 expanded grad, which is what `out.sum().backward()`
produces, before the batch check is ever reached, and CUDA only dodges it because its early
return sits above that check. Real losses produce dense grads, so this does not bite in
practice.

Test Plan:
buck2 test fbcode//mode/opt-amd-gpu -c fbcode.rocm_arch=mi300 fbcode//caffe2/test:test_transformers_cuda -- --regex 'test_fused_attention_empty_batch'

https://www.internalfb.com/intern/testinfra/testrun/14073749030862000

Reviewed By: kqfu

Differential Revision: D118590692

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @jataylo @hongxiayang @naromero77amd @pragupta @jerrymannil @xinyazhang

### CODE REVIEW

**Review by sylvesterkaczmarek — COMMENTED** [review]
The empty-batch early exit still sits below the stride-0 `dout` validation, so the standard `out.sum().backward()` path continues to fail; the regression avoids it with `randn_like(out)`. CUDA returns before that check. Could we move the empty-batch return earlier and cover a scalar-reduction backward?

**Review by kqfu — APPROVED** [review]

**Comment by rmiao** [comment]
> The empty-batch early exit still sits below the stride-0 `dout` validation, so the standard `out.sum().backward()` path continues to fail; the regression avoids it with `randn_like(out)`. CUDA returns before that check. Could we move the empty-batch return earlier and cover a scalar-reduction backward?

Thanks for the suggestion. I have fixed by making the check vacuous instead of relocating the return. A empty-batch  has no meaningful last-dim stride, so nothing is weakened and the behavior matches CUDA.

**Comment by rmiao** [comment]
Hi @sylvesterkaczmarek, could you please help review this change again? Thanks!
